packages/compiler-cli/src/transformers/patch_alias_reference_resolution.ts

```diff
@@ -60,32 +63,47 @@ interface EmitResolver {
  * `emitDecoratorMetadata` flag is enabled. TypeScript basically surfaces the same problem and
  * solves it conceptually the same way, but obviously doesn't need to access an `@internal` API.
  *
+ * The set that is returned by this function is meant to be filled with import declaration nodes
+ * that have been referenced in a value-position by the transform, such the the installed patch can
+ * ensure that those import declarations are not elided.
+ *
  * See below. Note that this uses sourcegraph as the TypeScript checker file doesn't display on
  * Github.
  * https://sourcegraph.com/github.com/microsoft/TypeScript@3eaa7c65f6f076a08a5f7f1946fd0df7c7430259/-/blob/src/compiler/checker.ts#L31219-31257
  */
-export function patchAliasReferenceResolutionOrDie(
-    context: ts.TransformationContext, referencedAliases: Set<ts.Declaration>): void {
+export function loadAliasReferenceResolutionPatchOrDie(context: ts.TransformationContext):
+    Set<ts.Declaration> {
   // If the `getEmitResolver` method is not available, TS most likely changed the
   // internal structure of the transformation context. We will abort gracefully.
   if (!isTransformationContextWithEmitResolver(context)) {
     throwIncompatibleTransformationContextError();
-    return;
   }
   const emitResolver = context.getEmitResolver();
+
+  // The emit resolver may have been patched already, in which case we return the set of referenced
+  // aliases that was created when the patch was first applied.
+  // See https://github.com/angular/angular/issues/40276.
+  const existingReferencedAliases = emitResolver[patchedReferencesAliasesSymbol];
+  if (existingReferencedAliases !== undefined) {
+    return existingReferencedAliases;
+  }
+
+  const referencedAliases = new Set<ts.Declaration>();
+
   const originalReferenceResolution = emitResolver.isReferencedAliasDeclaration;
   // If the emit resolver does not have a function called `isReferencedAliasDeclaration`, then
   // we abort gracefully as most likely TS changed the internal structure of the emit resolver.
   if (originalReferenceResolution === undefined) {
     throwIncompatibleTransformationContextError();
-    return;
   }
   emitResolver.isReferencedAliasDeclaration = function(node, ...args) {
```

Comment:
We should also change this function declaration to avoid the unnecessary variable argument:

```suggestion
  emitResolver.isReferencedAliasDeclaration = function(node, checkChildren) {
```

Reply:
I kept it as is on purpose, to avoid potential issues with TS updates which could introduce additional parameters without us being aware (as this is internal API, we don't have accurate type-checking of this call).

Reply:
In that case I would propose that we change the `EmitResolver` interface above to have the same signature.

Reply:
Good idea, done.